File: src/main/java/com/chattriggers/ctjs/minecraft/objects/display/Display.java
package com.chattriggers.ctjs.minecraft.objects.display;

import jdk.nashorn.api.scripting.ScriptObjectMirror;
import lombok.Getter;
import lombok.Setter;
import lombok.experimental.Accessors;
import lombok.experimental.Tolerate;

import java.util.ArrayList;

@Accessors(chain = true)
public class Display {
    /**
     * -- GETTER --
     * Gets the list of lines currently displayed
     *
     * @return An ArrayList of {@link DisplayLine} objects
     *
     * -- SETTER --
     * Sets the list of lines currently displayed
     * @param lines An ArrayList of {@link DisplayLine} objects
     * @return The Display object to allow for method chaining
     */
    @Getter
    @Setter
    private ArrayList<DisplayLine> lines;

    /**
     * -- GETTER --
     * Gets the x position of the Display
     *
     * @return The x position of the Display
     *
     * -- SETTER --
     * Sets the x position of the Display
     * @param renderX The new x location
     * @return The Display object to allow for method chaining
     */
    @Getter
    @Setter
    private float renderX;

    /**
     * -- GETTER --
     * Gets the y position of the Display
     *
     * @return The y position of the Display
     *
     * -- SETTER --
     * Sets the y position of the display
     * @param renderY The new y location
     * @return The Display object to allow for method chaining
     */
    @Getter
    @Setter
    private float renderY;

    /**
     * -- GETTER --
     * Gets the render status of the display (boolean)
     *
     * @return True if the Display is set to render
     *
     * -- SETTER --
     * Sets whether or not the Display should render
     * @param shouldRender True if the display should render
     * @return The Display object to allow for method chaining
     */
    @Getter
    @Setter
    private boolean shouldRender;

    /**
     * -- GETTER --
     * Gets a Display's background type
     *
     * @return The Display's background
     *
     * -- SETTER --
     * Sets a Display's background type.
     * @param background The new Display background
     * @return The Display object to allow for method chaining
     * @see DisplayHandler.Background
     * @see DisplayHandler.Background
     */
    @Getter
    @Setter
    private DisplayHandler.Background background;

    /**
     * -- GETTER --
     * Gets the Display's background color
     *
     * @return The Display's background color
     *
     * -- SETTER --
     * Sets the Display's background color
     * @param backgroundColor The integer color of the background
     * @return The Display object to allow for method chaining
     */
    @Getter
    @Setter
    private int backgroundColor;

    /**
     * -- GETTER --
     * Gets the Display's text color
     *
     * @return The Display's text color
     *
     * -- SETTER --
     * Sets the Display's text color
     * @param textColor The integer color of the text
     * @return The Display object to allow for method chaining
     */
    @Getter
    @Setter
    private int textColor;

    /**
     * -- GETTER --
     * Gets the Display's text alignment
     *
     * @return The Display's text alignment
     *
     * -- SETTER --
     * Sets the Displays text alignment
     * @param align The new Display alignment
     * @return The Display object to allow for method chaining
     * @see DisplayHandler.Align
     * @see DisplayHandler.Align
     */
    @Getter
    @Setter
    private DisplayHandler.Align align;

    /**
     * -- GETTER --
     * Gets the Display's line order
     *
     * @return The Display's line order
     *
     * -- SETTER --
     * Sets the display's line order
     * @param order The new Display line order
     * @return The Display object to allow for method chaining
     * @see DisplayHandler.Order
     * @see DisplayHandler.Order
     */
    @Getter
    @Setter
    private DisplayHandler.Order order;

    /**
     * -- GETTER --
     * Gets the minimum width of the Display
     *
     * @return The minimum width of the Display
     *
     * -- SETTER --
     * Sets the minimum width of the Display
     * @param minWidth The new minimum Display width
     * @return The Display object to allow for method chaining
     */
    @Getter
    @Setter
    private int minWidth;

    /**
     * -- GETTER --
     * Gets the width of the Display
     *
     * @return The width of the Display
     */
    @Getter
    private float width;

    /**
     * -- GETTER --
     *
     * @return the height of the Display
     */
    @Getter
    private float height;

    public Display() {
        this.lines = new ArrayList<>();

        this.shouldRender = true;
        this.renderX = 0f;
        this.renderY = 0f;

        this.background = DisplayHandler.Background.NONE;
        this.backgroundColor = 0x50000000;
        this.textColor = 0xffffffff;
        this.align = DisplayHandler.Align.LEFT;
        this.order = DisplayHandler.Order.DOWN;

        this.minWidth = 0;

        DisplayHandler.getInstance().registerDisplay(this);
    }

    public Display(ScriptObjectMirror config) {
        this.lines = new ArrayList<>();

        this.shouldRender = (boolean) config.getOrDefault("shouldRender", true);
        this.renderX = (int) config.getOrDefault("renderX", 0);
        this.renderY = (int) config.getOrDefault("renderY", 0);

        this.background = (DisplayHandler.Background) config.getOrDefault("background", DisplayHandler.Background.NONE);
        this.backgroundColor = (int) config.getOrDefault("backgroundColor", 0x50000000);
        this.textColor = (int) config.getOrDefault("textColor", 0xffffffff);
        this.align = (DisplayHandler.Align) config.getOrDefault("align", DisplayHandler.Align.LEFT);
        this.order = (DisplayHandler.Order) config.getOrDefault("order", DisplayHandler.Order.DOWN);

        this.minWidth = (int) config.getOrDefault("minWidth", 0);

        DisplayHandler.getInstance().registerDisplay(this);
    }

    /**
     * Sets a Display's background type using string input.
     * <p>
     *     Input can be:<br>
     *         &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;"full"<br>
     *         &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;"per line"<br>
     * </p>
     *
     * @param background the type of background
     * @return the Display to allow for method chaining
     */
    @Tolerate
    public Display setBackground(String background) {
        this.background = DisplayHandler.Background.getBackgroundByName(background);
        return this;
    }

    /**
     * Sets a Display's text alignment using string input.
     * <p>
     *     Input can be:<br>
     *         &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;"right"<br>
     *         &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;"left"<br>
     *         &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;"center"
     * </p>
     *
     * @param align the type of alignment
     * @return the Display to allow for method chaining
     */
    @Tolerate
    public Display setAlign(String align) {
        this.align = DisplayHandler.Align.getAlignByName(align);
        return this;
    }

    /**
     * Sets a Display's line order using string input.
     * <p>
     *     Input can be:<br>
     *         &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;"up"<br>
     *         &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;"down"
     * </p>
     *
     * @param order the order of lines
     * @return the display to allow method chaining
     */
    @Tolerate
    public Display setOrder(String order) {
        this.order = DisplayHandler.Order.getOrderByName(order);
        return this;
    }


    /**
     * Sets a Display line to a string.
     *
     * @param lineNumber the line number to set (0 based)
     * @param line       the string to set the line to
     * @return the Display to allow for method chaining
     */
    public Display setLine(int lineNumber, String line) {
        return setLine(lineNumber, new DisplayLine(line));
    }

    /**
     * Sets a Display line to a DisplayLine
     *
     * @param lineNumber the line number to set (0 based)
     * @param line       the DisplayLine to set the line to
     * @return the Display to allow for method chaining
     */
    public Display setLine(int lineNumber, DisplayLine line) {
        while (this.lines.size() - 1 < lineNumber)
            this.lines.add(new DisplayLine(""));

        this.lines.set(lineNumber, line);

        return this;
    }

    /**
     * Gets a DisplayLine from a line in a display.
     *
     * @param lineNumber the line number to get
     * @return the DisplayLine in the line of Display
     */
    public DisplayLine getLine(int lineNumber) {
        try {
            return lines.get(lineNumber);
        } catch (Exception exception) {
            return new DisplayLine("");
        }
    }

    /**
     * Adds multiple lines to the end of a Display.<br>
     * Arguments can be either a String or a DisplayLine.
     *     
     * @param lines the lines to add
     * @return the Display to allow for method chaining
     */
    public Display addLines(Object... lines) {
        for (Object line : lines) {
            if (line instanceof String) {
                this.lines.add(new DisplayLine((String) line));
            } else if (line instanceof DisplayLine) {
                this.lines.add((DisplayLine) line);
            }
        }
        
        return this;
    }

    /**
     * Adds one line to a Display.<br>
     * Arguments can be either a String or a DisplayLine.
     * 
     * @param line the line to add
     * @return the Display to allow for method chaining
     */
    public Display addLine(Object line) {
        return addLines(line);
    }

    /**
     * Clears all the lines in the Display.
     *
     * @return the Display to allow for method chaining
     */
    public Display clearLines() {
        lines.clear();
        return this;
    }

    /**
     * Set the X and Y render position of the Display.
     *
     * @param renderX the x coordinate
     * @param renderY the y coordinate
     * @return the Display to allow for method chaining
     */
    public Display setRenderLoc(float renderX, float renderY) {
        this.renderX = renderX;
        this.renderY = renderY;
        return this;
    }


    // Renders the display on to the player's screen.
    public void render() {
        if (!shouldRender) return;

        int maxWidth = this.minWidth;
        for (DisplayLine line : lines) {
            if (line.getTextWidth() > maxWidth)
                maxWidth = (int) Math.ceil(line.getTextWidth());
        }

        this.width = maxWidth;

        float i = 0;

        for (DisplayLine line : lines) {
            drawLine(line, this.renderX, this.renderY + (i * 10), maxWidth);

            if (order == DisplayHandler.Order.DOWN)
                i += line.getText().getScale();
            else if (order == DisplayHandler.Order.UP)
                i -= line.getText().getScale();
        }

        this.height = Math.abs(i);
    }

    // helper method to draw line with align
    private void drawLine(DisplayLine line, float x, float y, int maxWidth) {
        if (this.align == DisplayHandler.Align.LEFT)
            line.drawLeft(x, y, maxWidth, this.background, this.backgroundColor, this.textColor);
        else if (this.align == DisplayHandler.Align.RIGHT)
            line.drawRight(x, y, maxWidth, this.background, this.backgroundColor, this.textColor);
        else if (this.align == DisplayHandler.Align.CENTER)
            line.drawCenter(x, y, maxWidth, this.background, this.backgroundColor, this.textColor);
    }
}

File: src/main/kotlin/com/chattriggers/ctjs/minecraft/objects/display/Display.kt
package com.chattriggers.ctjs.minecraft.objects.display

import com.chattriggers.ctjs.minecraft.libs.renderer.Renderer
import com.chattriggers.ctjs.utils.kotlin.External
import com.chattriggers.ctjs.utils.kotlin.NotAbstract
import org.mozilla.javascript.NativeObject
import java.util.concurrent.CopyOnWriteArrayList

@External
@NotAbstract
abstract class Display {
    private var lines = CopyOnWriteArrayList<DisplayLine>()

    private var renderX = 0f
    private var renderY = 0f
    private var shouldRender = true
    private var order = DisplayHandler.Order.DOWN

    private var backgroundColor: Long = 0x50000000
    private var textColor: Long = 0xffffffff
    private var background = DisplayHandler.Background.NONE
    private var align = DisplayHandler.Align.LEFT

    private var minWidth = 0f
    private var width = 0f
    private var height = 0f

    internal var registerType = DisplayHandler.RegisterType.RENDER_OVERLAY

    constructor() {
        @Suppress("LeakingThis")
        DisplayHandler.registerDisplay(this)
    }

    constructor(config: NativeObject?) {
        shouldRender = config.getOption("shouldRender", true).toBoolean()
        renderX = config.getOption("renderX", 0).toFloat()
        renderY = config.getOption("renderY", 0).toFloat()

        setBackgroundColor(config.getOption("backgroundColor", 0x50000000).toLong())
        setBackground(config.getOption("background", DisplayHandler.Background.NONE))
        setTextColor(config.getOption("textColor", 0xffffffff).toLong())
        setAlign(config.getOption("align", DisplayHandler.Align.LEFT))
        setOrder(config.getOption("order", DisplayHandler.Order.DOWN))

        minWidth = config.getOption("minWidth", 0f).toFloat()

        setRegisterType(config.getOption("registerType", DisplayHandler.RegisterType.RENDER_OVERLAY))

        @Suppress("LeakingThis")
        DisplayHandler.registerDisplay(this)
    }

    private fun NativeObject?.getOption(key: String, default: Any): String {
        return (this?.get(key) ?: default).toString()
    }

    fun getBackgroundColor(): Long = backgroundColor

    fun setBackgroundColor(backgroundColor: Long) = apply {
        this.backgroundColor = backgroundColor
    }

    fun getTextColor(): Long = textColor

    fun setTextColor(textColor: Long) = apply {
        this.textColor = textColor
    }

    fun getBackground(): DisplayHandler.Background = background

    fun setBackground(background: Any) = apply {
        this.background = when (background) {
            is String -> DisplayHandler.Background.valueOf(background.uppercase().replace(" ", "_"))
            is DisplayHandler.Background -> background
            else -> DisplayHandler.Background.NONE
        }
    }

    fun getAlign(): DisplayHandler.Align = align

    fun setAlign(align: Any) = apply {
        this.align = when (align) {
            is String -> DisplayHandler.Align.valueOf(align.uppercase())
            is DisplayHandler.Align -> align
            else -> DisplayHandler.Align.LEFT
        }
    }

    fun getOrder(): DisplayHandler.Order = order

    fun setOrder(order: Any) = apply {
        this.order = when (order) {
            is String -> DisplayHandler.Order.valueOf(order.uppercase())
            is DisplayHandler.Order -> order
            else -> DisplayHandler.Order.DOWN
        }
    }

    fun setLine(index: Int, line: Any) = apply {
        while (lines.size - 1 < index)
            lines.add(createDisplayLine(""))

        lines[index] = when (line) {
            is String -> createDisplayLine(line)
            is DisplayLine -> line
            else -> createDisplayLine("")
        }
    }

    fun getLine(index: Int): DisplayLine = lines[index]

    fun getLines(): List<DisplayLine> = lines

    fun setLines(lines: MutableList<DisplayLine>) = apply {
        this.lines = CopyOnWriteArrayList(lines)
    }

    @JvmOverloads
    fun addLine(index: Int = -1, line: Any) = apply {
        val toAdd = when (line) {
            is String -> createDisplayLine(line)
            is DisplayLine -> line
            else -> createDisplayLine("")
        }

        if (index == -1) {
            lines.add(toAdd)
        } else lines.add(index, toAdd)
    }

    fun addLines(vararg lines: Any) = apply {
        this.lines.addAll(lines.map {
            when (it) {
                is String -> createDisplayLine(it)
                is DisplayLine -> it
                else -> createDisplayLine("")
            }
        })
    }

    fun clearLines() = apply {
        lines.clear()
    }

    fun getRenderX(): Float = renderX

    fun setRenderX(renderX: Float) = apply {
        this.renderX = renderX
    }

    fun getRenderY(): Float = renderY

    fun setRenderY(renderY: Float) = apply {
        this.renderY = renderY
    }

    fun setRenderLoc(renderX: Float, renderY: Float) = apply {
        this.renderX = renderX
        this.renderY = renderY
    }

    fun getShouldRender(): Boolean = shouldRender

    fun setShouldRender(shouldRender: Boolean) = apply {
        this.shouldRender = shouldRender
        lines.forEach { it.shouldRender = shouldRender }
    }

    fun getWidth(): Float = width

    fun getHeight(): Float = height

    fun getMinWidth(): Float = minWidth

    fun setMinWidth(minWidth: Float) = apply {
        this.minWidth = minWidth
    }

    /**
     * Gets the type of register the display will render under.
     *
     * The returned value will be a DisplayHandler.RegisterType
     *      renderOverlayEvent: render overlay
     *      postGuiRenderEvent: post gui render
     *
     * @return the register type
     */
    fun getRegisterType() : DisplayHandler.RegisterType = registerType


    /**
     * Sets the type of register the display will render under.
     *
     * Possible input values are:.
     *      renderOverlayEvent: render overlay
     *      postGuiRenderEvent: post gui render
     */
    fun setRegisterType(registerType: Any) = apply {
        this.registerType = when (registerType) {
            is String -> DisplayHandler.RegisterType.valueOf(registerType.uppercase().replace(" ", "_"))
            is DisplayHandler.RegisterType -> registerType
            else -> DisplayHandler.RegisterType.RENDER_OVERLAY
        }
    }

    fun render() {
        if (!shouldRender)
            return

        width = lines.maxOfOrNull { it.getTextWidth() }?.coerceAtLeast(minWidth) ?: minWidth

        var i = 0f
        lines.forEach {
            it.draw(renderX, renderY + i, width, background, backgroundColor, textColor, align)

            when (order) {
                DisplayHandler.Order.DOWN -> i += it.getText().getHeight()
                DisplayHandler.Order.UP -> i -= it.getText().getHeight()
            }
        }

        height = i
    }

    internal abstract fun createDisplayLine(text: String): DisplayLine

    override fun toString() =
        "Display{" +
                "shouldRender=$shouldRender, registerType=$registerType, " +
                "renderX=$renderX, renderY=$renderY, " +
                "background=$background, backgroundColor=$backgroundColor, " +
                "textColor=$textColor, align=$align, order=$order, " +
                "minWidth=$minWidth, width=$width, height=$height, " +
                "lines=$lines" +
                "}"

}
